### PR TITLE
gh-116397: Move the _PyGC_ClearAllFreeLists to the safe point

### PR DESCRIPTION
**Pull request by corona10**
* Issue: gh-116397

### CODE REVIEW

File: Lib/test/test_sys.py [hunk 1]
```diff
@@ -1046,7 +1046,7 @@ def test_getallocatedblocks(self):
         self.assertLessEqual(b, a)
         gc.collect()
         c = sys.getallocatedblocks()
-        self.assertIn(c, range(b - 50, b + 50))
+        self.assertLessEqual(c, b)
```
Comment (corona10) [inline Lib/test/test_sys.py]: There is the change for free-threaded build.
But IIUC, `c` is always less equal than `b`, but not sure why the range comparation was applied that was not used when comparing `a` and `b`.

First introduced at: https://github.com/python/cpython/commit/f9d0b1256fd5a9ae90fa8e8418bd450ec6b7f5f2

**Review by colesbury — COMMENTED** [review]
We should run the refleak buildbots on this PR. If the refleak tracking or allocated blocks tests are a problem, we may also want to clear the current thread's free-list again after the GC (i.e., where we previously were calling `_PyGC_ClearAllFreeLists()`). That should be safe and I expect will fix the tests.

File: Python/gc_free_threading.c [hunk 2]
```diff
@@ -1160,8 +1162,11 @@ gc_collect_main(PyThreadState *tstate, int generation, _PyGC_Reason reason)
             n+m, n, d);
     }
 
-    // Clear free lists in all threads
-    _PyGC_ClearAllFreeLists(interp);
+    // Clear the current thread's free-list again.
+    HEAD_LOCK(&_PyRuntime);
```
Comment (colesbury) [inline Python/gc_free_threading.c]: This doesn't need a `HEAD_LOCK()` because we aren't traversing the linked list of threads.

Comment (corona10) [reply Python/gc_free_threading.c]: Ah you are right :)

**Review by colesbury — APPROVED** [review]

**Pull request merged**